feat: cache project info

blurfx opened the pull request:
<!--  Thanks for sending a pull request! -->

**What this PR does / why we need it**:

Cache ProjectInfo to improve response time.

For caching I used the existing LRU cache implementation because it seems there is no external cache layer like redis.

In cluster mode, each node has its own cache, so there is still a possibility of query requests for same project info.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #568

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation**:

<!--
This section can be blank if this pull request does not require a release note.

Please use the following format for linking documentation:
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

**Checklist**:
- [x] Added relevant tests or not required
- [x] Didn't break anything

hackerwins commented on server/rpc/interceptors/context.go:
```diff
@@ -34,15 +39,24 @@ import (
 	"github.com/yorkie-team/yorkie/server/rpc/metadata"
 )
 
+const projectInfoCacheSize = 1
```
It looks like the size of the cache is 1, so the cache only holds one project in a time. I wonder if it is intended.

blurfx replied:
Yes it's intended, I understood that a each project can have one ProjectInfo, isn't it?

blurfx replied:
No it's not intended. I had it set to 256, but I think I accidentally changed it. 

and here's something I'd like to discuss. how much CacheSize and TTL are appropriate?

I was going to set the CacheSize to 256 and the TTL to 1 hour, but what do you think?

hackerwins replied:
I also think 256 and 1 hour are suitable for the current situation. This is because yorkie.dev doesn't have many projects yet and project updates don't affect requests from clients.

hackerwins reviewed (commented):
Thanks for your contribution. I left a small comment.

blurfx commented:
@hackerwins 

Integration test `auth_webhook_test.go` fails after the change.

It passes when testing only the `Selected method authorization webhook test` test case, but fails when testing the entire test suite and the `authorization webhook test` test is run first.

I tried to fix it, but there's something I don't understand. If using an invalid token when activating the client below, shouldn't we get an Unauthenticated error?

https://github.com/yorkie-team/yorkie/blob/dba636f8af25b9da8e941a3b12330da865d0440c/test/integration/auth_webhook_test.go#L164-L173

hackerwins reviewed (commented):
Thank you for applying my requests.

The reason for CI failure

The purpose of the `Selected method authorization webhook test` is to verify that the interceptor only checks authentication and authorization of specific methods. In this test case, the selected methods are `AttachDocument` and `WatchDocuments`. As a result, the `ActivateClient` method passes even with an invalid token.

After implementing a cache for projects, the project information can be cached from the previous test. Consequently, the project info may not be immediately updated in the current test. So the test failed.

We should take into cache eviction options and modify the test to wait for a certain period until it gets evicted by TTL.

blurfx commented:
in 576e3c1, I moved the `ProjectInfoCacheSize` and `ProjectInfoCacheTTL` to the config. therefore, I added the default values and also added them to the sample config.

now test backend has 5 seconds `ProjectInfoCacheTTL`, and I modified the existing test to do `time.Sleep(5 * time.Seconds)`.

krapie reviewed (approved):
> In cluster mode, each node has its own cache, so there is still a possibility of query requests for same project info.

Since we are using sharded cluster mode, I think we will have a meaningful cache hit rate for querying the same project info.

LGTM!

hackerwins reviewed (commented):
Thanks for applying my requests. I left one more question about the cache key.

> Since we are using sharded cluster mode, I think we will have a meaningful cache hit rate for querying the same project info.

We have two types of APIs: project-wide APIs and document-wide APIs. For document-wide APIs, requests are shared based on the document key, causing requests to be evenly distributed across nodes for a project. So it seems likely that caches for all projects will exist on each node.

blurfx commented on server/rpc/interceptors/context.go:
```diff
@@ -146,15 +155,19 @@ func (i *ContextInterceptor) buildContext(ctx context.Context) (context.Context,
 		md.Authorization = authorization[0]
 	}
 	ctx = metadata.With(ctx, md)
+	cacheKey := fmt.Sprintf("%s:%s", md.APIKey, md.Authorization)
```
I used the authorization header by misunderstanding. so yes, cache key should be just apikey without auth header. I’ll fix it.
